EIPS/eip-7792.md

```diff
@@ -0,0 +1,126 @@
+
+def track_log(evm: Evm, entry: LogEntry) -> None:
+    entry_root = entry.hash_tree_root()
+
+    # Allow verification via `address` filter
+    key = keccak256(abi.encode(entry.log.address, LOG_ADDRESS_STORAGE_SLOT))
+    accumulate_log(evm, entry_root, key)
+
+    for topic in entry.log.topics:
+        # Allow verification via `topics` filter
+        key = keccak256(abi.encode(topic, LOG_TOPICS_STORAGE_SLOT))
+        accumulate_log(evm, entry_root, key)
+
+        # Allow verification via combined `address` + `topics` filter
+        key = keccak256(abi.encode(entry.log.address, topic))
+        key = keccak256(abi.encode(key, LOG_ADDRESS_TOPICS_STORAGE_SLOT))
+        accumulate_log(evm, entry_root, key)
+```
+
+### JSON-RPC API
+
+The `eth_getLogs` response format is extended to include:
+
+- `blockTimestamp`: `QUANTITY` - The timestamp field of the block referred to by `blockHash`
+
+### Verification
+
+For `eth_getLogs(address, topics, fromBlock, toBlock)`, the response data can be verified for correctness and completion by obtaining:
+
+1. `fromBlock` and `toBlock` block headers (validated against their known hashes)
+2. `fromBlock`'s `parentBlock` header (validated against `fromBlock.parentHash`)
+3. Historical log accumulator at `parentBlock` based on given filters (validated with `eth_getProof`)
+4. Log accumulator at `toBlock` based on given filters (validated with `eth_getProof`)
+
+Starting from the historical log accumulator from (3), each response entry is applied to it in a way compatible with `accumulate_log` above. If the log accumulator ends up matching the value from (4), the response data is correct and `LogEntry` derived from it can be trusted.
+
+## Rationale
+
+Making the `eth_getLogs` response verifiable adds the necessary security attributes to enable wallets to transition away from relying on trusted data providers, ultimately improving the wallet's privacy guarantees as it is no longer subject to the privacy policy of any given provider.
+
+### Gas cost
+
+The gas cost produced by this scheme is significantly higher than what `LOG#` opcodes produce as of Prague, primarily due to the additional `SLOAD` / `SSTORE` requirement and the double cost of `SHA256` opcodes compared to `KECCAK256` opcodes. The gas cost increases outweigh the savings from dropping logs blooms.
+
+If the mechanism turns out to be prohibitively expensive even when optimized, it may be necessary to move the log accumulators to a separate optimized data structure (not in `state_root`), or to an out-of-protocol zk system. Even then, the gas cost for logs should still reflect the actual overall cost to update a typical out-of-protocol accumulator to deter against log spamming.
+
+## Backwards Compatibility
+
+It is still possible to process `eth_getLogs` responses from trusted servers as is, without verifying them. Client applications with strict response validation may need to be updated to allow the additional `blockTimestamp` field.
```

Comment:
This change needs a fork since this writes as a system-call like to a system contract and is thus not backwards-compatible.

Reply:
Yes, a fork is needed to introduce such a scheme if done inside a contract. Backwards compatibility addresses existing application logic, which continues to work as is because the extra writes to the `LOG_CONTRACT_ADDRESS` storage slots do not interfere with existing code. As in, a soft fork is sufficient.

If a zk system is used to store the hash chain state, it may not even need a fork.

Note this is in draft stage at this time, not scheduled for inclusion.